experimental/dds/tree2/src/feature-libraries/default-field-kinds/optionalField.ts

```diff
@@ -33,21 +35,117 @@ import { populateChildModifications } from "../deltaUtils";
+	}
+
+	public keys(): Iterable<ChangeId> {
+		const changeIds: ChangeId[] = [];
+		for (const [localId, nestedMap] of this.nestedMapData) {
+			if (localId === "self") {
+				changeIds.push("self");
+			} else {
+				for (const [revisionTag, _] of nestedMap) {
+					changeIds.push(
+						revisionTag === undefined
+							? { localId }
+							: { localId, revision: revisionTag },
+					);
+				}
+			}
+		}
+
+		return changeIds;
+	}
+	public values(): Iterable<T> {
+		return this.nestedMapData.values();
+	}
+	public entries(): Iterable<[ChangeId, T]> {
+		const entries: [ChangeId, T][] = [];
+		for (const changeId of this.keys()) {
+			if (changeId === "self") {
+				const entry = this.nestedMapData.tryGet("self", undefined);
+				assert(entry !== undefined, "Entry should not be undefined when iterating keys.");
+				entries.push(["self", entry]);
+			} else {
+				const entry = this.nestedMapData.tryGet(changeId.localId, changeId.revision);
+				assert(entry !== undefined, "Entry should not be undefined when iterating keys.");
+				entries.push([changeId, entry]);
+			}
+		}
+
+		return entries;
+	}
+	public get size(): number {
+		return this.nestedMapData.size;
+	}
+}
+
 export const optionalChangeRebaser: FieldChangeRebaser<OptionalChangeset> = {
 	compose: (
 		changes: TaggedChange<OptionalChangeset>[],
 		composeChild: NodeChangeComposer,
 	): OptionalChangeset => {
+		const perChildChanges = new ChildChangeMap<TaggedChange<NodeChangeset>[]>();
+		const addChildChange = (id: ChangeId, ...changeList: TaggedChange<NodeChangeset>[]) => {
+			const existingChanges = perChildChanges.get(id);
+			if (existingChanges !== undefined) {
+				existingChanges.push(...changeList);
+			} else {
+				perChildChanges.set(id, [...changeList]);
+			}
+		};
+
 		let fieldChange: Mutable<OptionalFieldChange> | undefined;
-		const origNodeChange: TaggedChange<NodeChangeset>[] = [];
-		const newNodeChanges: TaggedChange<NodeChangeset>[] = [];
+		let currentChildNodeChanges: TaggedChange<NodeChangeset>[] = [];
 		for (const { change, revision } of changes) {
-			if (change.deletedBy === undefined && change.childChange !== undefined) {
-				const taggedChange = tagChange(change.childChange, revision);
-				if (fieldChange === undefined) {
-					origNodeChange.push(taggedChange);
-				} else {
-					newNodeChanges.push(taggedChange);
+			const { childChanges } = change;
+			if (childChanges !== undefined) {
+				for (const [childId, childChange] of childChanges) {
+					const taggedChildChange = tagChange(childChange, revision);
+					if (childId === "self") {
+						// childChange refers to the node that existed at the start of `change`,
+						// Thus in the composition, it should be referred to by whatever deletes that node in the future, which is what
+						// currentChildNodeChanges tracks (note that since we handle childChanges before fieldChange, this applies to
+						// if the current change does delete that node).
```

Comment:
"this applies to if the current change does delete that node"
I'm not sure what this is supposed to say.

Reply:
it's probably not super necessary, mostly just pointing out that the ordering of handling the childChanges within a change and handling the field change is relevant due to this assumption.

The parenthetical is specifically calling out compositions of changes which have a fieldChange as well as childChanges to the node that they delete.